From 292a31b5d634bfd844cd646d2b327a5d62d851b0 Mon Sep 17 00:00:00 2001
From: "Bartosik, Lukasz" <Lukasz.Bartosik@cavium.com>
Date: Tue, 27 Nov 2018 14:32:39 +0300
Subject: [PATCH 0773/1051] crypto: cpt - implement workaround for HW issue

Implement software workaround for HW issue related to CPT.
http://mcbuggin.caveonetworks.com/bug/35946.

Signed-off-by: Lukasz Bartosik <lukasz.bartosik@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/crypto/cavium/cpt/9x/cpt9x_lf_main.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/crypto/cavium/cpt/9x/cpt9x_lf_main.c b/drivers/crypto/cavium/cpt/9x/cpt9x_lf_main.c
index 28628a2b34e3..c534a618cf0f 100644
--- a/drivers/crypto/cavium/cpt/9x/cpt9x_lf_main.c
+++ b/drivers/crypto/cavium/cpt/9x/cpt9x_lf_main.c
@@ -418,15 +418,20 @@ static irqreturn_t cptlf_misc_intr_handler(int irq, void *cptlf)
 static irqreturn_t cptlf_done_intr_handler(int irq, void *cptlf)
 {
 	struct cptlf_info *lf = (struct cptlf_info *) cptlf;
+	union cptx_vqx_done_wait done_wait;
 	int irq_cnt;
 
 	/* Read the number of completed requests */
 	irq_cnt = cptlf_read_done_cnt(lf);
 	if (irq_cnt) {
-
+		done_wait.u = cpt_read64(lf->lfs->reg_base, BLKADDR_CPT0,
+					 lf->slot, CPT_LF_DONE_WAIT);
 		/* Acknowledge the number of completed requests */
 		cpt_write64(lf->lfs->reg_base, BLKADDR_CPT0, lf->slot,
 			    CPT_LF_DONE_ACK, irq_cnt);
+
+		cpt_write64(lf->lfs->reg_base, BLKADDR_CPT0, lf->slot,
+			    CPT_LF_DONE_WAIT, done_wait.u);
 		if (unlikely(!lf->wqe)) {
 			dev_err(&lf->lfs->pdev->dev, "No work for LF %d\n",
 				lf->slot);
-- 
2.17.1

